src/core/createConsent.js

```diff
@@ -0,0 +1,264 @@
+/*
+Copyright 2019 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import { cookieJar, defer, getNamespacedCookieName, values } from "../utils";
+import {
+  OPTIN_COOKIE_KEY,
+  OPTOUT_COOKIE_KEY
+} from "../constants/cookieDetails";
+import parseConsentCookie from "./parseConsentCookie";
+
+const getPersistedConsent = cookieName => {
+  const cookieValue = cookieJar.get(cookieName);
+  if (cookieValue) {
+    return parseConsentCookie(cookieValue);
+  }
+  return undefined;
+};
+
+const areAllPurposesTrue = preferencesByPurpose => {
+  return (
+    preferencesByPurpose && values(preferencesByPurpose).every(value => value)
+  );
+};
+
+const areAnyPurposesTrue = preferencesByPurpose => {
+  return values(preferencesByPurpose).some(value => value);
+};
+
+const OPTED_OUT_OF_ALL_PURPOSES = "The user opted out of all purposes.";
+const OPTED_INTO_NO_PURPOSES = "The user opted into no purposes.";
+
+const createConsentState = ({
+  optInEnabled,
+  optInCookieName,
+  optOutCookieName,
+  onChange
+}) => {
+  let optInPreferencesByPurpose;
+  let optOutPreferencesByPurpose;
+
+  const getStatus = () => {
+    // We look for any purpose being true rather than just looking for the
+    // "general" purpose being true because it gives us some flexibility in the
+    // future. If the edge decides to split the "general" purpose into
+    // individual purposes, like "tracking" and "personalization", this code
+    // would conservatively consider the user opted out of everything if either
+    // "tracking" or "personalization" were set to true. This buys us time
+    // to release a new version of Alloy that appropriately behaves according
+    // to the more granular purposes and for customers to upgrade. In the
+    // case of opt-in, this is also why we look for all purposes being true
+    // rather than just the "general" purpose being true.
```

Comment:
This is worth pointing out. Right now the opt-in and opt-out cookie values look like `GENERAL=true` or `GENERAL=false`, but that's just because we don't have more granular purposes. It's likely that `GENERAL` goes away and we start using more granular purposes.